suite/tests/client-interface/drx-scattergather.dll.c

```diff
@@ -160,21 +161,23 @@ event_bb_app2app(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
     }
     CHECK((scatter_gather_present IF_X64(&&expanded)) || (expansion_ok && !expanded),
           "drx_expand_scatter_gather() bad OUT values");
-    if (one_time_mask_clobber_test) {
+    if ((!translating && one_time_mask_clobber_test) ||
+        (translating && one_time_mask_clobber_translating)) {
         for (instr = instrlist_first(bb); instr != NULL; instr = instr_get_next(instr)) {
             if (instr_get_opcode(instr) == OP_kandnw) {
-                instr_t *search_app_instr = instr;
-                while (!instr_is_app(search_app_instr)) {
-                    search_app_instr = instr_get_next(search_app_instr);
-                }
                 /* We've found the clobber case of the scatter/gather sequence that
                  * clobbers the k0 mask register. Then we're inserting a ud2 app
-                 * instruction right after it, so we will SIGILL and can test the value in
-                 * the app's signal handler.
+                 * instruction right after it, so we will SIGILL and can test the
+                 * value in the app's signal handler.
                  */
-                instrlist_postinsert(bb, instr,
-                                     INSTR_XL8(INSTR_CREATE_ud2a(drcontext),
-                                               instr_get_app_pc(search_app_instr)));
+                instrlist_postinsert(
+                    bb, instr,
+                    INSTR_XL8(INSTR_CREATE_ud2a(drcontext),
+                              instr_get_app_pc(instr_get_next_app(instr))));
```

Comment:
There's guaranteed to be a next app instr?  I.e., this won't be NULL?

Reply:
Yes, it's guaranteed by the test (first mask update, there are always about 15 * N more app instructions afterwards)

Reply:
It seems worth an assert or comment.

Reply:
Done.